### ISSUE TITLE
Shell command line completion for hledger

### ISSUE THREAD

**Issue by schoettl** [simonmichael/hledger#945]
Shell completion of the command line, triggered by the tab key, would be nice.

I started to write one for Bash: https://asciinema.org/a/PdV2PzIU9oDQg1K5FjAX9n3vL

It completes:

- commands and generic options
- comand-specific options
- account names (but not yet for files named by `--file`)
- filenames for options that take a filename as argument

The Bash completion script is generated (GNU make) by parsing output of `hledger`, `hledger -h`, and `hledger <cmd> -h`. The script also uses `hledger accounts` for account name completion. I propose that the Makefile is *not* run at every built but rather manually when the CLI changes.

Can I provide a pull request when it's ready? What is a good place to put the scripts in? Folder `shell-completion`? Where does the code go to install the completion script on a Linux system? The main Makefile?

But first, I have to get it ready for deployment. Currently it only runs in a specific directory and requires some external files.

**Comment by simonmichael** [comment]
> I think it's not too bad to maintain as long as the usage output of hledger does not change fundamentally. Because the important parts of the completion script are auto-generated with simple rules.

It doesn't change often, but it has changed and will likely change again.. 

I find command line handling pretty complicated, with all the permutations of common options, command-specific options, builtin commands and addon commands.

> Do you know some platform package maintainers? E.g. Felix Yan for Arch Linux. I guess he creates the installation scripts and can also install the completion script. It should be one line like of code.
> 

You should be able to find them by following the platform links on the download page.

**Comment by simonmichael** [comment]
We now provide [bash shell completions support](https://github.com/simonmichael/hledger/tree/master/shell-completion thanks to @schoettl . Future work for anyone interested: extend to other popular shells, make more discoverable to hledger users, help packagers to add it to their packages.

**Comment by schoettl** [comment]
I plan to work on your last to points during the next time...

**Comment by simonmichael** [comment]
In case it's useful in future: [cmdargs](https://hackage.haskell.org/package/cmdargs-0.10.15/docs/System-Console-CmdArgs-Explicit.html) has the ability to generate bash completion info from our flag definitions.

**Comment by simonmichael** [comment]
Yes I imagine using it only once and saving the completions, not running hledger every time. I expect we would need to add support for it to expose this feature.
